File: Dynamic Programming/Hard/LET_no1289_Minimum Falling Path Sum II.py
# Given a square grid of integers arr, a falling path with non-zero shifts
# is a choice of exactly one element from each row of arr, such that no two
# elements chosen in adjacent rows are in the same column.
# Return the minimum sum of a falling path with non-zero shifts.

class Solution:
    def minFallingPathSum(self, arr: List[List[int]]) -> int:

        for i in range(1, len(arr)):
            r = sorted(arr[i - 1])
            for j in range(len(arr[0])):
                arr[i][j] += r[1] if r[0] == arr[i - 1][j] else r[0]

        return min(arr[-1])

File: Dynamic Programming/Hard/LET_no1289_Minimum Falling Path Sum II.cpp
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        for (int i = 1; i < m; i++) {
            auto p = helper(grid[i - 1]);
            for (int j = 0; j < n; j++)
                grid[i][j] += (p.first == grid[i - 1][j] ? p.second : p.first);
        }

        auto p = helper(grid[m - 1]);
        return p.first;
    }

    pair<int, int> helper(vector<int>& arr) {
        int min = INT_MAX, min2 = INT_MAX;
        for (int num : arr) {
            if (num < min) {
                min2 = min;
                min = num;
            }
            else if (num < min2)
                min2 = num;
        }
        
        return {min, min2};
    }
};